src/REL/d/a/obj/d_a_obj_paint.cpp

```diff
@@ -1,9 +1,137 @@
 #include "d/a/obj/d_a_obj_paint.h"
 
+#include "d/a/d_a_player.h"
+#include "d/a/obj/d_a_obj_base.h"
+#include "d/col/bg/d_bg_s.h"
+#include "d/snd/d_snd_wzsound.h"
+#include "f/f_base.h"
+#include "m/m_mtx.h"
+#include "m/m_quat.h"
+#include "m/m_vec.h"
+#include "nw4r/g3d/res/g3d_resfile.h"
+#include "nw4r/g3d/res/g3d_resmdl.h"
+#include "rvl/MTX/mtx.h"
+#include "toBeSorted/arc_managers/oarc_manager.h"
+
 SPECIAL_ACTOR_PROFILE(OBJ_PAINT, dAcOpaint_c, fProfile::OBJ_PAINT, 0x163, 0, 0);
 
 STATE_DEFINE(dAcOpaint_c, Wait);
 
+mAng dAcOpaint_c::rotationZRelated0 = mAng::d2s(35.0025f);
+mAng dAcOpaint_c::rotationZRelated1 = mAng::d2s(40.0015f);
+
+bool dAcOpaint_c::createHeap() {
+    const char *oarcName = getOarcName();
+    const char *modelName = getModelName();
+    void *data = getOarcResFile(oarcName);
+    mResFile = nw4r::g3d::ResFile(data);
+    nw4r::g3d::ResMdl mdl = mResFile.GetResMdl(modelName);
+    TRY_CREATE(mMdl.create(mdl, &mAllocator, 0x120, 1, nullptr));
+
+    void *dzb = OarcManager::GetInstance()->getDzbFromArc(oarcName, modelName);
+    void *plc = OarcManager::GetInstance()->getPlcFromArc(oarcName, modelName);
+    TRY_CREATE(!mBgW.Set((cBgD_t *)dzb, (PLC *)plc, 1, &mWorldMtx, &mScale));
+    return SUCCEEDED;
+}
+
+int dAcOpaint_c::create() {
+    if (getFromParams(0, 3) == 0) {
+        mSubtype = 0;
+    } else {
+        mSubtype = 1;
+    }
+    dAcObjBase_c::updateMatrix();
+    CREATE_ALLOCATOR(dAcOpaint_c);
+    mMdl.setLocalMtx(mWorldMtx);
+    dBgS::GetInstance()->Regist(&mBgW, this);
+    field_0x5AA = false;
+    mStateMgr.changeState(StateID_Wait);
+    return SUCCEEDED;
+}
+
+int dAcOpaint_c::doDelete() {
+    return SUCCEEDED;
+}
+
+int dAcOpaint_c::actorExecute() {
+    mStateMgr.executeState();
+    updateMatrix();
+    mBgW.Move();
+    return SUCCEEDED;
+}
+
+int dAcOpaint_c::draw() {
+    drawModelType1(&mMdl);
+    static mQuat_c rot(0.0f, 0.0f, 0.0f, 100.0f);
+    fn_8002edb0(mShadow, mMdl, &rot, -1, -1, 0.0f);
+    return SUCCEEDED;
+}
+
 void dAcOpaint_c::initializeState_Wait() {}
-void dAcOpaint_c::executeState_Wait() {}
+
+void dAcOpaint_c::executeState_Wait() {
+    s32 _a = 0;
+    if (dAcPy_c::GetLink() != nullptr && dAcPy_c::GetLink()->checkFlags0x350(0x2000)) {
+        mVec3_c deltaPosition = dAcPy_c::GetLink()->mPosition - mPosition;
+        f32 distance = deltaPosition.absXZ();
+        if (1000.f - distance > 0.f) {
+            if (distance < _a + 700.f && !field_0x5AA) {
+                if (mSubtype == 0) {
+                    mRotationZRelated = rotationZRelated0;
+                } else {
+                    mRotationZRelated = rotationZRelated1;
+                }
+                field_0x5AA = true;
+                startSound(SE_Paint_LOOSE);
+            }
+            field_0x5A8 = 3;
+            mPosition.y += 10.f;
+
+            if (field_0x5AA) {
+                mRotationZOffset = mAng(0x100);
+            }
+        }
+    }
```

Comment:
This appears to be responsable for 'swaying' the painting in the case of link hitting the wall?

field_0x5A8 is the time to offset the paintings position
field_0x5AA is also the boolean to indicate this already happened. It